subworkflows/local/genotype.nf

```diff
@@ -0,0 +1,373 @@
+            }
+
+        // TODO Do we want to provide SNP capture bed file to Freebayes? It would then genotype only on those positions.
+        // NOTE: dbsnp is not used by Freebayes, but we need to provide it to the module anyway, to ensure correct cardinality of the fasta channel within the BCFTOOLS_STATS channel operations.
+        //     i.e. to keep the definition of the ch_fasta_for_multimap channel consistent regardless of genotyper, so the `combine -> multiMap` in lines 327-328 work.
+        ch_fasta_for_multimap = ch_fasta_plus
+            // Because dbsnp is optional, the channel can be [[],[]]. remainder:true will output both the empty list and the fasta_plus channel with an added 'null'.
+            .join( ch_dbsnp, remainder:true ) // [ [ref_meta], fasta, fai, dict, dbsnp ]
+            // Also filter out the empty list dbsnp (meta == [])
+            .filter { it[0] != [] }
+            // convert added null dbsnp into an empty list
+            .map {
+                ref_meta, fasta, fai, dict, dbsnp ->
+                def final_dbsnp = dbsnp != null ? dbsnp : []
+                [ ref_meta, fasta, fai, dict, final_dbsnp ]
+            }
+            .map {
+            // Prepend a new meta that contains the meta.id value as the new_meta.reference attribute
+                WorkflowEager.addNewMetaFromAttributes( it, "id" , "reference" , false )
+            } // RESULT: [ [combination_meta], [ref_meta], fasta, fai, dict, dbsnp ]
+
+        ch_input_for_freebayes = ch_bams_for_multimap
+            .combine( ch_fasta_for_multimap , by:0 )
+            .multiMap {
+                ignore_me, meta, bam, bai, ref_meta, fasta, fai, dict, dbsnp ->
+                    bam:   [ meta, bam , bai, [], [], [] ] // No second bam, second bai, or regions-BED file
+                    fasta: [ ref_meta, fasta ]
+                    fai:   [ ref_meta, fai ]
+            }
+
+        // TODO: Should the vcfs be indexed with bcftools index? VCFs from HC are indexed.
+        FREEBAYES(
+            ch_input_for_freebayes.bam,
+            ch_input_for_freebayes.fasta,
+            ch_input_for_freebayes.fai,
+            [ [], [] ], // No samples file
+            [ [], [] ], // No populations file
+            [ [], [] ]  // No CNV file
+            )
+        ch_freebayes_genotypes = FREEBAYES.out.vcf
+        ch_versions            = ch_versions.mix( FREEBAYES.out.versions.first() )
+    }
+
+    if ( params.genotyping_tool == 'angsd' ) {
+        // TODO no module for angsd genotyping yet
+    }
+
+    // Run BCFTOOLS_STATS on output from GATK UG, HC and Freebayes
+    if ( !params.skip_bcftools_stats && ( params.genotyping_tool == 'hc' || params.genotyping_tool == 'ug' || params.genotyping_tool == 'freebayes' ) ) {
+        ch_bcftools_input= ch_gatk_unifiedgenotyper_genotypes
+            .mix( ch_gatk_haplotypecaller_genotypes )
+            .mix( ch_freebayes_genotypes )
```

Comment:
What about a  comment output channel for all the genotope files (or at least for each file type: vcf, eigenstratocaster etc.)?  Woudl also  mean you have less empty channels on initialisaiton